### ISSUE TITLE
clush sometimes block when trying to handle stdin after ssh exits

### ISSUE THREAD

**Issue by martinetd** [cea-hpc/clustershell#279]
When the remote command ends early and clush has read something to transmit from stdin, the input loop will use msg() which waits until it is actually sent and never return from stdin_port.msg(buf)

Using send_once would work if there is little input, but there's the risk of filling the msg queue then (defaults to 100), it's actually easy to fill with something like:

    yes | clush -w foo echo ok

I can't think of anything better than starting the input loop thread as daemonic, so python will exit even if we're blocked there -- e.g.
```patch
diff --git a/lib/ClusterShell/CLI/Clush.py b/lib/ClusterShell/CLI/Clush.py
index ab7e85d..361be7d 100755
--- a/lib/ClusterShell/CLI/Clush.py
+++ b/lib/ClusterShell/CLI/Clush.py
@@ -625,7 +625,9 @@ def bind_stdin(worker, display):
     # Launch a dedicated thread to read stdin in blocking mode. Indeed stdin
     # can be a file, so we cannot use a WorkerSimple here as polling on file
     # may result in different behaviors depending on selected engine.
-    threading.Thread(None, _stdin_thread_start, args=(port, display)).start()
+    stdin_thread = threading.Thread(None, _stdin_thread_start, args=(port, display))
+    stdin_thread.setDaemon(True)
+    stdin_thread.start()
 
 def run_command(task, cmd, ns, timeout, display, remote):
     """
```
(didn't see any option to declare the thread as daemonic?)

A friend of this bug is trying to write into the ssh fd after ssh exits, which will raise a broken pipe error with this backtrace (bypassing clush exception handler):

```traceback
Traceback (most recent call last):
  File "/usr/bin/clush", line 10, in <module>
    main()
  File "/path/to/clustershell/lib/ClusterShell/CLI/Clush.py", line 1045, in main
    options.remote != 'no')
  File "/path/to/clustershell/lib/ClusterShell/CLI/Clush.py", line 667, in run_command
    task.resume()
  File "/path/to/clustershell/lib/ClusterShell/Task.py", line 799, in resume
    self._resume()
  File "/path/to/clustershell/lib/ClusterShell/Task.py", line 762, in _resume
    self._run(self.timeout)
  File "/path/to/clustershell/lib/ClusterShell/Task.py", line 407, in _run
    self._engine.run(timeout)
  File "/path/to/clustershell/lib/ClusterShell/Engine/Engine.py", line 690, in run
    self.runloop(timeout)
  File "/path/to/clustershell/lib/ClusterShell/Engine/EPoll.py", line 165, in runloop
    client._handle_read(sname)
  File "/path/to/clustershell/lib/ClusterShell/Worker/EngineClient.py", line 510, in _handle_read
    self.eh.ev_msg(self, pmsg.get())
  File "/path/to/clustershell/lib/ClusterShell/CLI/Clush.py", line 87, in ev_msg
    self.master_worker.write(msg)
  File "/path/to/clustershell/lib/ClusterShell/Worker/Exec.py", line 357, in write
    client._write(sname, buf)
  File "/path/to/clustershell/lib/ClusterShell/Worker/EngineClient.py", line 397, in _write
    self._handle_write(sname)
  File "/path/to/clustershell/lib/ClusterShell/Worker/EngineClient.py", line 320, in _handle_write
    wcnt = os.write(wfile.fd, wfile.wbuf)
OSError: [Errno 32] Broken pipe
```

We could hide this error by wrapping the write call, but I think we should leave this as is because it can tell the user something went wrong... Most people don't want to lose lines, and it's not obvious that a "while read foo; do clush ...; done" loop won't work unless they use --nostdin (it doesn't work with ssh either unless you close stdin with e.g. 0<&-)

**Comment by degremont** [comment]
The second part of this ticket is a duplicate of #196

**Comment by thiell** [comment]
Thanks @martinetd for the detailed bug report! Let's handle the second part in #196 as I'm not sure this will be fixed in 1.7.x. For the first part of the problem, we need to fix it in 1.7.1. According to some stackoverflow threads I just read, it looks like the right way to do it. We also could maybe use atexit and then kill the thread, but I guess this makes sense to make this thread daemonic. Please note that the java-like API `thread.setDaemon(True)` is needed for python 2.4 support. But `thread.daemon = True` is the new 2.6+ way to mark a thread daemonic (so for now we should use `setDaemon()` as you propose and we'll change it later).

@martinetd: are you ok to submit a change to gerrit for this part as you already did the patch? :)
